lib/ansible/plugins/strategy/__init__.py

```diff
@@ -302,6 +308,96 @@ def add_tqm_variables(self, vars, play):
         vars['ansible_current_hosts'] = self.get_hosts_remaining(play)
         vars['ansible_failed_hosts'] = self.get_failed_hosts(play)
 
+    def _short_circuit_fork(self, host, task, task_vars, templar):
+        self._short_circuit_fork_conditonal(host, task, task_vars, templar)
+        self._short_circuit_fork_include(host, task, task_vars, templar)
+
+    def _short_circuit_fork_conditonal(self, host, task, task_vars, templar):
+        if not task.loop and not task.loop_with:
+            tr = None
+            try:
+                conditional = task.evaluate_conditional(templar, task_vars)
+            except AnsibleError as e:
+                tr = TaskResult(
+                    host=host.name,
+                    task=task._uuid,
+                    return_data={
+                        'failed': True,
+                        'msg': to_text(e),
+                        'exception': traceback.format_exc(),
+                    },
+                )
+            else:
+                if not conditional:
+                    tr = TaskResult(
+                        host=host.name,
+                        task=task._uuid,
+                        return_data={
+                            'changed': False,
+                            'skipped': True,
+                            'skip_reason': 'Conditional result was False',
+                        },
+                    )
+            if tr:
+                raise ForkShortCircuit(tr)
+
+    def _short_circuit_fork_include(self, host, task, task_vars, templar):
+        vars_copy = task_vars.copy()
```

Comment:
Wouldn't it be better to do the copy in the if statement. Not sure how expensive this operation is but I would have thought `task_vars` could be quite large if the host has a lot of vars/facts set on it.